src/systems/rtti.rs

```diff
@@ -0,0 +1,312 @@
+        };
+        out.into()
+    }
+
+    #[inline]
+    pub fn get_global_functions(&self) -> Vec<Function> {
+        let mut out = Array::default();
+        unsafe { (self.vft().base.get_global_functions)(&(*self.0)._base, &mut out.0 as *mut _) };
+        out.into()
+    }
+
+    #[inline]
+    pub fn get_class_functions(&self) -> Vec<Function> {
+        let mut out = Array::default();
+        unsafe { (self.vft().base.get_class_functions)(&(*self.0)._base, &mut out.0 as *mut _) };
+        out.into()
+    }
+
+    /// retrieve base class and its inheritors, optionally including abstract classes
+    /// while allowing custom filter.
+    #[inline]
+    pub fn get_classes(
+        &self,
+        base: &Class,
+        mut filter: Option<&mut dyn FnMut(&Class) -> bool>,
+        include_abstract: bool,
+    ) -> Vec<Class> {
+        let mut classes = Array::<*const red::CClass>::default();
+        unsafe {
+            (self.vft().base.get_classes)(
+                &(*self.0)._base,
+                &base.0,
+                &mut classes.0 as *mut _,
+                None,
+                include_abstract,
+            )
+        };
+
+        if filter.is_none() {
+            return classes.into();
+        }
+
+        let mut out = Vec::<Class>::with_capacity(classes.0.size as usize);
+        for class in &classes {
+            if !filter.as_mut().unwrap()(class.as_ref()) {
+                continue;
+            }
+            out.push((*class).into());
+        }
+        out
+    }
+
+    /// retrieve derived classes, omitting base in the output.
+    #[inline]
+    pub fn get_derived_classes(&self, base: &Class) -> Vec<Class> {
+        let mut out = Array::<*const red::CClass>::default();
+        unsafe {
+            (self.vft().base.get_derived_classes)(&(*self.0)._base, &base.0, &mut out.0 as *mut _)
+        };
+        out.into()
+    }
+
+    #[inline]
+    pub fn get_class_by_script_name(&self, name: CName) -> Option<&Class> {
+        let ty = unsafe { (self.vft().base.get_class_by_script_name)(&(*self.0)._base, name.0) };
+        if ty.is_null() {
+            return None;
+        }
+        Some(unsafe { &*ty.cast::<Class>() })
+    }
+
+    #[inline]
+    pub fn get_enum_by_script_name(&self, name: CName) -> Option<&Enum> {
+        let ty = unsafe { (self.vft().base.get_enum_by_script_name)(&(*self.0)._base, name.0) };
+        if ty.is_null() {
+            return None;
+        }
+        Some(unsafe { &*ty.cast::<Enum>() })
+    }
+
+    #[inline]
+    fn vft(&self) -> &RTTISystemVft {
+        unsafe { &*((*self.0)._base.vtable_ as *const RTTISystemVft) }
+    }
+}
+
+#[repr(C)]
+struct RTTISystemVft {
+    base: IRTTISystemVft,
+}
+
+#[repr(C)]
+struct IRTTISystemVft {
+    pub get_type: unsafe extern "fastcall" fn(
+        this: *const red::IRTTISystem,
+        name: red::CName,
+    ) -> *const red::CBaseRTTIType,
```

Comment:
`this` should always be mutable, I'd also define it to use our wrappers for convenience
```suggestion
    get_type: unsafe extern "fastcall" fn(
        this: *mut IRTTISystem,
        name: CName,
    ) -> *const Type
```

Reply:
But then we're gonna have to do `self as *const _ as *mut _` in every function? Or maybe I'm missing something.

Reply:
please see my comment above, since both are related.

Reply:
We can change it to *const for methods we know don't need mutability